pkg/monitor/operator.go

```diff
@@ -0,0 +1,369 @@
+						}
+						if i != last {
+							items.Items[last] = *item
+							last++
+						}
+					}
+					items.Items = items.Items[:last]
+				}
+				return items, err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				w, err := client.CoreV1().Pods("").Watch(options)
+				if err == nil {
+					w = watch.Filter(w, func(in watch.Event) (watch.Event, bool) {
+						return in, filterToOperatorPods(in.Object)
+					})
+				}
+				return w, err
+			},
+		}),
+		&corev1.Pod{},
+		10*time.Minute,
+		nil,
+	)
+
+	m.AddSampler(func(now time.Time) []*Condition {
+		var conditions []*Condition
+		for _, obj := range podInformer.GetStore().List() {
+			pod, ok := obj.(*corev1.Pod)
+			if !ok {
+				continue
+			}
+
+			podReady, err := utils.PodRunningReady(pod)
+			if !podReady || err != nil {
+				conditions = append(conditions, &Condition{
+					Level:   Warning,
+					Locator: locatePod(pod),
+					Message: "skipped, pod is not Running",
+				})
+				continue
+			}
+			if len(pod.Spec.Containers) == 0 {
+				conditions = append(conditions, &Condition{
+					Level:   Warning,
+					Locator: locatePod(pod),
+					Message: "skipped, no containers found",
+				})
+				continue
+			}
+
+			metrics, err := getPodMetrics(config, pod, podURLGetter)
+			if err != nil {
+				switch err.(type) {
+				case *url.Error:
+					if strings.Contains(err.Error(), "EOF") {
+						conditions = append(conditions, &Condition{
+							Level:   Warning,
+							Locator: locatePod(pod),
+							Message: fmt.Sprintf("/metrics endpoint not available"),
+						})
+					}
+				default:
+					conditions = append(conditions, &Condition{
+						Level:   Error,
+						Locator: locatePod(pod),
+						Message: fmt.Sprintf("error retrieving pod metrics: %v", err),
+					})
+				}
+				continue
+			}
+
+			metricGroup, ok := metrics["rest_client_requests_total"]
+			if !ok {
+				conditions = append(conditions, &Condition{
+					Level:   Error,
+					Locator: locatePod(pod),
+					Message: fmt.Sprintf("error: failed to find counter: %q", "rest_client_requests_total"),
+				})
+				continue
+			}
+
+			procStartTime, ok := metrics["process_start_time_seconds"]
+			if !ok || len(procStartTime.Metric) == 0 {
+				conditions = append(conditions, &Condition{
+					Level:   Error,
+					Locator: locatePod(pod),
+					Message: fmt.Sprintf("error: failed to find metric: %q", "process_start_time_seconds"),
+				})
+				continue
+			}
+			procStartTimeSeconds := procStartTime.Metric[0].GetGauge().GetValue()
+			totalProcTimeSeconds := time.Now().Unix() - int64(procStartTimeSeconds)
+
+			totalRequestCount := float64(0)
+			for _, metric := range metricGroup.Metric {
+				totalRequestCount += metric.Counter.GetValue()
+			}
+
+			qpsLimit := maxQPSAllowed
+			if customLimit, ok := perComponentNamespaceMaxQPSAllowed[pod.Namespace]; ok {
+				qpsLimit = customLimit
+			}
+			qps := totalRequestCount / float64(totalProcTimeSeconds)
+
+			if qps > qpsLimit {
```

Comment:
Why is this not an e2e test?

Please explain why your test can't be equivalent to prometheus tests that grab qps metrics and check them, something exactly like `should report telemetry if a cloud.openshift.com token is present`

Reply:
@smarterclayton Thanks, opened https://github.com/openshift/origin/pull/21852 to attempt to implement this